### PR TITLE
Tool Validator Node

### PR DESCRIPTION
**Pull request by hinthornw**
Basically, for generating validated structured output in a conversational setting.

Example:

```python
from typing import Literal

from langchain_anthropic import ChatAnthropic
from langchain_core.pydantic_v1 import BaseModel, validator

from langgraph.graph import END, START, MessageGraph
from langgraph.prebuilt import ValidationNode


class SelectNumber(BaseModel):
    a: int

    @validator("a")
    def a_must_be_meaningful(cls, v):
        if v != 37:
            raise ValueError("Only 37 is allowed")
        return v


builder = MessageGraph()
llm = ChatAnthropic(model="claude-3-haiku-20240307").bind_tools([SelectNumber])
builder.add_node("model", llm)
builder.add_node("validation", ValidationNode([SelectNumber]))
builder.add_edge(START, "model")


def should_validate(state: list) -> Literal["validation", "__end__"]:
    if state[-1].tool_calls:
        return "validation"
    return END


builder.add_conditional_edges("model", should_validate)


def should_reprompt(state: list) -> Literal["model", "__end__"]:
    for msg in state[::-1]:
        # None of the tool calls were errors
        if msg.type == "ai":
            return END
        if msg.additional_kwargs.get("is_error"):
            return "model"
    return END


builder.add_conditional_edges("validation", should_reprompt)


def get_ai_message(state: list):
    for msg in state[::-1]:
        if msg.type == "ai":
            return msg
    raise ValueError("No AI message found")


graph = builder.compile()
res = graph.invoke(("user", "Select a number, any number"))
res[-2].pretty_print()
```

### CODE REVIEW

File: langgraph/prebuilt/__init__.py [hunk 1]
```diff
@@ -1,9 +1,51 @@
...
+if TYPE_CHECKING:
+    from langgraph.prebuilt import chat_agent_executor
+    from langgraph.prebuilt.agent_executor import create_agent_executor
+    from langgraph.prebuilt.chat_agent_executor import create_react_agent
+    from langgraph.prebuilt.tool_executor import ToolExecutor, ToolInvocation
+    from langgraph.prebuilt.tool_node import ToolNode, tools_condition
+    from langgraph.prebuilt.tool_validator import ValidationNode
+
+
+def __getattr__(name: str) -> Any:
```
Comment (nfcampos) [inline langgraph/prebuilt/__init__.py:14]: Why do we need this?

File: langgraph/prebuilt/tool_validator.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        ...             return END
+        ...         if msg.additional_kwargs.get("is_error"):
+        ...             return "model"
+        ...     return END
+        ...
+        ...
+        >>> builder.add_conditional_edges("validation", should_reprompt)
+        ...
+        ...
+        >>> def get_ai_message(state: list):
```
Comment (nfcampos) [inline langgraph/prebuilt/tool_validator.py:125]: Is this being used?

File: langgraph/prebuilt/tool_validator.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        ...
+        >>> builder.add_conditional_edges("model", should_validate)
+        ...
+        ...
+        >>> def should_reprompt(state: list) -> Literal["model", "__end__"]:
+        ...     for msg in state[::-1]:
+        ...         # None of the tool calls were errors
+        ...         if msg.type == "ai":
+        ...             return END
+        ...         if msg.additional_kwargs.get("is_error"):
```
Comment (nfcampos) [inline langgraph/prebuilt/tool_validator.py:117]: A little sad we can't just return dict content in messages

File: langgraph/prebuilt/__init__.py (hunk 1 above)
Comment (hinthornw) [reply langgraph/prebuilt/__init__.py:14]: ha i'll remove apologies - meant to have that as a separate pr

File: langgraph/prebuilt/tool_validator.py (hunk 2 above)
Comment (hinthornw) [reply langgraph/prebuilt/tool_validator.py:125]: ha no

File: langgraph/prebuilt/tool_validator.py (hunk 3 above)
Comment (hinthornw) [reply langgraph/prebuilt/tool_validator.py:117]: ya ... :(

**Pull request merged**